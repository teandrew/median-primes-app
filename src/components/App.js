import React, { Component } from "react";
import "./App.css";
import Form from "react-bootstrap/Form";
import Button from "react-bootstrap/Button";
import Col from "react-bootstrap/Col";
import axios from "axios";

import Result from "./Result";

class App extends Component {
  state = {
    error: "",
    limit: "",
    isLoading: false,
    medianPrimes: []
  };

  handleChange = event => {
    this.setState({
      limit: event.target.value
    });
  };

  handleSubmit = event => {
    event.preventDefault();
    const limit = this.state.limit;

    if (limit === "") {
      return;
    }

    if (isNaN(limit)) {
      return this.setState({
        error: "Limit must be a number",
        medianPrimes: []
      });
    }

    return this.getMedianPrimes(limit);
  };

  getMedianPrimes = limit => {
    this.setState({ isLoading: true });
    return axios
      .get(`http://localhost:4040/median-prime/${limit}`)
      .then(response => {
        this.setState({
          error: "",
          isLoading: false,
          medianPrimes: response.data
        });
      })
      .catch(({ response }) => {
        const errorMessage = response
          ? response.data.message
          : "Something went wrong";
        this.setState({
          error: errorMessage,
          medianPrimes: [],
          isLoading: false
        });
      });
  };

  render() {
    return (
      <div className="App">
        <div style={{ marginBottom: "20px" }}>
          <h1>Find the median primes</h1>
        </div>

        <div className="form-container">
          <Form onSubmit={this.handleSubmit}>
            <Form.Row>
              <Col>
                <Form.Control
                  id="limit"
                  name="limit"
                  autoComplete="off"
                  style={{ marginBottom: "10px" }}
                  onChange={this.handleChange}
                  placeholder="Enter a number"
                  value={this.state.limit}
                />
                {this.state.error && !this.state.isLoading && (
                  <div id="error" style={{ color: "red" }}>
                    <p>{this.state.error}</p>
                  </div>
                )}
              </Col>
              <Col xs={2}>
                <Button id="submit-btn" type="submit" variant="primary">
                  Get
                </Button>
              </Col>
            </Form.Row>
          </Form>
          {this.state.isLoading && <div className="loader" />}
          {this.state.medianPrimes.length > 0 && !this.state.isLoading && (
            <Result medianPrimes={this.state.medianPrimes} />
          )}
        </div>
      </div>
    );
  }
}

export default App;
