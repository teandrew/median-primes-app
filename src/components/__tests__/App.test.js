import React from "react";
import App from "../App";
import axios from "axios";
import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";

jest.mock("axios");

describe("App Component", () => {
  it("renders", () => {
    const wrapper = shallow(<App />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should respond to change event and change state", () => {
    const wrapper = shallow(<App />);

    wrapper.find("#limit").simulate("change", {
      target: { name: "limit", value: "10" }
    });

    expect(wrapper.state("limit")).toEqual("10");
  });

  it("renders Results when medianPrimes change", () => {
    const wrapper = mount(<App />);

    wrapper.setState({
      medianPrimes: [2]
    });

    expect(wrapper.find("p").text()).toEqual("The median prime(s): 2");
  });

  it("should change median primes state", () => {
    const wrapper = shallow(<App />);

    axios.get.mockImplementation(() => Promise.resolve({ data: [2, 3] }));

    return wrapper
      .instance()
      .getMedianPrimes(3)
      .then(() => {
        expect(wrapper.state("medianPrimes")).toEqual([2, 3]);
      });
  });

  it("should set error message on error", () => {
    const wrapper = shallow(<App />);

    axios.get.mockImplementation(() =>
      Promise.reject({
        response: { data: { message: "Limit must be a number" } }
      })
    );

    return wrapper
      .instance()
      .getMedianPrimes("abcd")
      .then(() => {
        expect(wrapper.state("error")).toEqual("Limit must be a number");
      });
  });
});
