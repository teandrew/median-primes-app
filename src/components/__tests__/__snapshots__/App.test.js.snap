// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component renders 1`] = `
<div
  className="App"
>
  <div
    style={
      Object {
        "marginBottom": "20px",
      }
    }
  >
    <h1>
      Find the median primes
    </h1>
  </div>
  <div
    className="form-container"
  >
    <ForwardRef(Bootstrap(Form))
      onSubmit={[Function]}
    >
      <ForwardRef(Bootstrap(FormRow))>
        <ForwardRef(Bootstrap(Col))>
          <ForwardRef(ContextTransform)
            autoComplete="off"
            id="limit"
            name="limit"
            onChange={[Function]}
            placeholder="Enter a number"
            style={
              Object {
                "marginBottom": "10px",
              }
            }
            value=""
          />
        </ForwardRef(Bootstrap(Col))>
        <ForwardRef(Bootstrap(Col))
          xs={2}
        >
          <ForwardRef(Bootstrap(Button))
            id="submit-btn"
            type="submit"
            variant="primary"
          >
            Get
          </ForwardRef(Bootstrap(Button))>
        </ForwardRef(Bootstrap(Col))>
      </ForwardRef(Bootstrap(FormRow))>
    </ForwardRef(Bootstrap(Form))>
  </div>
</div>
`;
